Metric label proposal (#13975)
hashgraph/hedera-services: platform-sdk/docs/proposals/metric-labels/README.md

The metrics module should not be used by the end user.
Here the preferred solution is still that we forward our metrics to a monitoring system that the custom application code uses.

## Design & Architecture

This chapter describes the changes that are needed to add labels to the metrics module.

### About labels

Before we start with the implementation we need to define what labels are.
Labels are key-value pairs that are used to define, filter and group metrics.
Labels are used by monitoring systems like Grafana to create more dynamic dashboards.
By adding labels to the metrics module a single metric is now uniquely defined by its name and the labels.
For example, metrics that counts the number of requests can have a label called `method` that defines if the request was a `GET` or `POST` request.
That metrics are defined by the name "api_http_requests_total" and the label `method`.
By doing so the monitoring system can create a dashboard that shows the number of requests per method.
In the given example we would have 2 metric instances: one for `GET` and one for `POST`:

- Metric 1: `api_http_requests_total{method="GET"}`
- Metric 2: `api_http_requests_total{method="POST"}`

> While today the name of a metric is the unique identifier of a metric the name and the labels are the unique identifier of a metric when using labels.

Reviewer: I had to read this sentence a few times to get it. I suggest a minor rewording (immaterial):

```suggestion
While today the name of a metric is the unique identifier of a metric, the unique identifier will become the name and the labels once labels are added.
```

> By doing so both metrics can use the same name but are still unique as long as the labels are different.

Reviewer: Is this desirable? It seems quite confusing to me. Why not keep the unique identifier the same as before?

For example, it seems confusing to allow two metrics with the name `api_http_requests_total' with labels `method` and `handler` but allow another metric to be called `api_http_requests_total` with only the label `method`. To me, this should not be allowed.

We will use the same practices and definition as in prometheus regarding labels.

### Public API

The metrics module will be extended by a new data type called `Label`.
A label is a key-value pair that can be defined as a `record`:

```java
record Label(@NonNull String key, @Nullable String value) {}
``` 

> [!NOTE]  
The value of a label can be `null`.
For our prometheus endpoint a label with a `null` value will be handled as if the label would not exist.
This would allow us to define a label without a value and with a possible extension (see "Future Work") we define a concrete value for a label later.
Today best practice is to not use labels with a `null` value.

The `Metric` interface will be extended by a new method called `getLabels()` that returns a list of labels:

```java
interface Metric {
    
    //...
    
    @NonNull
    Set<Label> getLabels();
}
```

Like the name or category, the labels are defined when a metric is created and cannot be changed later.
Based on that the MetricConfig class will be extended by a new method called `withLabels`:

```java
abstract class MetricConfig {
    
    //...
    
    @NonNull

>     MetricConfig withLabels(@NonNull Label... labels) {...}

Reviewer: ```suggestion
    MetricConfig withLabels(@NonNull Label... label) {...}
```

    
    @NonNull
    MetricConfig withLabels(@NonNull Set<Label> labels) {...}
}
```

By adding labels the unique identifier of a metric is the name and the labels.
Based on that we need to modify the `Metrics` interface to use the labels as the unique identifier of a metric:

```java

interface Metrics {
    
    // old methods:
    // Object getValue(@NonNull String category, @NonNull String name);
    // void remove(@NonNull String category, @NonNull String name);
    
    @Nullable

>     Object getValue(@NonNull String category, @NonNull String name, @NonNull Set<Label> labels);

Reviewer: If we are going to include labels as part of the unique id, then it makes sense to me to have a`MetricId` record with these values in it. Then it is very clear what is part of a metric's unique identifier.

##### Examples

Here are some examples how to define a metric with labels:

```java
final Label transactionTypeLabel = new Label("transactionType", "fileUpload");
final Counter.Config config = new Counter.Config("transactionsCategory", "transactionCount").withLabels(transactionTypeLabel);

final Metrics metrics = ...;
final Counter counter = metrics.getOrCreate(config);

counter.increment();
```

In prometheus the metric will be defined like this:

```
transactionsCategory.transactionCount{transactionType="fileUpload"}
```

#### Snapshot API

The snapshot API will be extended to return the labels of a metric for a snapshot.
By doing so any monitoring system that uses the snapshot API can use the labels.
Today we only have 2 system that uses the snapshot API: the prometheus endpoint and the CSV exporter.
Since no class of the snapshot API is public today (based on the internal usage of `nodeId`) we plan to move parts
of the snapshot API to the public API.

### Migration

Next to extending the public API we need to migrate the current metrics to the new concept.

#### Migration of the `category`, `name`, and `nodeId`

> The `nodeId` information of a metric will be migrated to labels.

Reviewer 2: The proposal looks great overall. However, the only part that I cannot find here is how exactly it is proposed to add the new `nodeId` labels given that the `Label`'s value is a `String`. Would the code using this API (or the `MetricsFactory` implementation) just use the `NodeId.toString()`, or there will be a different mechanism to format a `NodeId` as a `String` for the purpose of putting it into a `Label`'s value ? Or perhaps the `Label` could be a generic type like `record Label<T>(String key, T value){}` and just use the `value.toString()` to get the actual string that you use as the value of the label? 

These are just some random options, and personally I don't have any preferences. But I'd expect this document to add some details about the `NodeId` label migration because currently it's unclear.

Author: Good point. Today the nodeID is already a label for prometheus. In our prometheus endpoint we therefore migrate a NodeId to a string. The same code will be used in future. The main difference is that this won't happen hidden in the prometheus endpoint code but when creating the `Metrics` instance for a platform.

Author: @anthony-swirldslabs do you want me to update the proposal with that information and create a new PR?

Reviewer 2: @hendrikebbers Thanks for clarifying. Given that this update wouldn't be material, I'd be okay with pushing it into this same PR. Another option would be to get this one merged, and then publish a tiny update PR. Or you can leave things as they are, given that you confirmed that you have a clear plan for this particular aspect. I'm good either way. I'll leave it up to you and other reviewers to decide.

Please feel free to resolve this conversation if you have enough approvals and would like to merge the current version of the proposal. Thanks!

Author: adding @poulok for feedback to get the process right

Reviewer: If you only need to add clarifying comments, i consider that immaterial and can be updated on this PR

Author: @anthony-swirldslabs please check my small change here: https://github.com/hashgraph/hedera-services/pull/13975/commits/37df54713e5572047b8647a22cd59e845a4a62c9

Reviewer 2: Looks great. Thank you!

# Metric labels

| Metadata           | Entities                                   | 
 |--------------------|--------------------------------------------|
 | Designers          | [@hendrikebbers](https://github.com/hendrikebbers), [@mxtartaglia-sl](https://github.com/mxtartaglia-sl) |

## Summary

This proposal describes a new feature for the metrics module.
The feature is called "metric labels" and allows to add labels to metrics.
Labels are key-value pairs that can be used to filter and group metrics.
Why this proposal uses the name "metric labels" (like prometheus does) other libraries (like micrometer or open telemetry) use the term tags or attributes.

## Motivation

Today the metrics module only supports the definition of metrics without adding custom labels.
While each metric has a hardcoded category, name, and nodeId no other custom information can be added.
This proposal suggests adding a new concept called "metric labels" to the metrics module and

> migrate the `category`, `name`, and `nodeId` information to labels.

Author: nothing. that sentence should have removed.

> migrate the `category`, `name`, and `nodeId` information to labels.

Author: done

While the category and name are defined as String values the nodeId is a class called `NodeId`.
By doing so the metrics module becomes more flexible and can be used 100% independent of the
platform module since it does not depend on the platform module's `nodeId` anymore.

## Goals

- Add a new concept called "metric labels" to the metrics module
- Migrate the `nodeId` information to a label
- Make the metrics module independent of the platform module
- Provide a public API to define labels for metrics
- Add labels to the snapshot API
- The output of the snapshot API should contain the labels
- The output of the snapshot API must be compatible with the current output (no breaking changes)
- Create better and more dynamic dashboards in Grafana by using labels

## Non-Goals

> Even with the new concept the metrics api should be only used in our (hashgraph) code.

Author: hashgraph is for me everything that we create in the Hedera ecosystem (inlcuding mirror node, ...). Maybe I need to find a better name.

> Even with the new concept the metrics api should be only used in our (hashgraph) code.

Author: done

While today the name of a metric is the unique identifier of a metric, the unique identifier will become the name and the labels once labels are added.
By doing so both metrics can use the same name but are still unique as long as the labels are different.
We will store measurements individually for each metric.
If you later want to know the combined measurement of all metrics with the name `api_http_requests_total` you can use the prometheus query language to sum up the values of all metrics with that name.
See https://prometheus.io/docs/prometheus/latest/querying/basics/#time-series-selectors for more information.

In prometheus a concrete metric is defined like this (see https://prometheus.io/docs/concepts/data_model/#notation):

```
api_http_requests_total{method="POST", handler="/messages"}
```

Here a metric is defined by a name (`api_http_requests_total`) and a set of labels (`method="POST"` and `handler="/messages"`).
In prometheus that defines the unique identifier of a metric.

When using labels to create dashboards in Grafana prometheus provides a rich query language to filter and group metrics: https://prometheus.io/docs/prometheus/latest/querying/examples/

> We will use the same practices and definition as in prometheus regarding labels.

Reviewer 3: Might be good to either link to those practice or call out some important ones to avoid.

Author: @poulok can I add the links without restart voting?

- https://prometheus.io/docs/practices/naming/#labels
 - https://prometheus.io/docs/practices/instrumentation/#use-labels

Author: done

### Public API

The metrics module will be extended by a new data type called `Label`.
A label is a key-value pair that can be defined as a `record`:

```java
record Label(@NonNull String key, @Nullable String value) {}
``` 

> [!NOTE]  
The value of a label can be `null`.
For our prometheus endpoint a label with a `null` value will be handled as if the label would not exist.
This would allow us to define a label without a value and with a possible extension (see "Future Work") we define a concrete value for a label later.

> Today best practice is to not use labels with a `null` value.

Author: you are right. Maybe it is nonnull for now.

> Today best practice is to not use labels with a `null` value.

Author: done

The `Metric` interface will be extended by a new method called `getLabels()` that returns a list of labels:

```java
interface Metric {
    
    //...
    
    @NonNull

>     Set<Label> getLabels();

Author: makes sense. Will add that

>     Set<Label> getLabels();

Author: done

We will have labels that are global for a `Metrics` instance.
The best example is the `nodeId` label.
We do not want to add the `nodeId` label to the creation of each metric.
Instead of that we want to add the `nodeId` label to the `Metrics` instance.
That should happen when a `Metrics` instance is created.
The creation of a `Metrics` instance is not public api today.
With the implementation of this proposal a `MetricsFactory` interface should be introduced that allows to create a `Metrics` instance.
The `MetricsFactory` interface will be defined like this:

```java
interface MetricsFactory {
    
    @NonNull
    default Metrics create() {
        return create(Set.of());
    }
    
    @NonNull
    Metrics create(@NonNull Set<Label> defaultLabels);
}
```

The definition of the api should follow our rules regarding services as defined at https://github.com/hashgraph/hedera-services/blob/develop/platform-sdk/docs/base/service-architecture/service-architecture.md

##### Examples

Here are some examples how to define a metric with labels:

```java
final Label transactionTypeLabel = new Label("transactionType", "fileUpload");

> final Counter.Config config = new Counter.Config("transactionsCategory", "transactionCount").withLabels(transactionTypeLabel);

Reviewer 4: I would prefer a solution that does not require me to create `Label` instances, e.g.:
```
final Counter.Config config = new Counter.Config("transactionsCategory", "transactionCount")
    .addLabel("transactionType", "fileUpload");
```

Author: Again both might be usefull. Will add that

### Migration

Next to extending the public API we need to migrate the current metrics to the new concept.

#### Migration of the `category`, `name`, and `nodeId`

The `nodeId` information of a metric will be migrated to labels.
Today we create a prometheus label out of the `nodeId` internally in our prometheus endpoint implementation.
The same conversion can be used to create a label for the `nodeId` in future but should be moved from the prometheus endpoint to the platform code.
The key of a metric won't change and still be defined as the combination of the `category` and `name`.
The unique identifier of the metric is the combination key and the labels.
That is a breaking change since today the key is the unique identifier of a metric.
Since we do not have an api today to add labels the migration should be minimalistic.

#### Migration of the prometheus endpoint

The `nodeId` of a platform is already added as a label today when using the prometheus endpoint.
For that endpoint we just need to remove the hard coded `nodeId` label and add a general algorithm to add labels.

#### Migration of the CSV exporter

The CSV exporter use the `nodeId` today to create the file name.

> Since the CSV exporter is only used by the platform module we can continue with that approach.

Reviewer 4: How are labels going to be used in the CSV-file?

Author: The nodeID label will be ignored and all other labels will be added to the name.

Author: But I assume it makes sense to make that more clear

final Counter counter = metrics.getOrCreate(config);

counter.increment();
```

Let's see how that metric can be used in a filter that filter transactions and counts the number of transactions of type `fileUpload`:

```java
class TransactionFilter {
    
    private final static Counter.Config config = new Counter.Config("transactionsCategory", "transactionCount");

    private final Metrics metrics;
    
    TransactionFilter(Metrics metrics) {
        this.metrics = metrics;
    }

    void handle(Transaction transaction) {
        if(transaction.getType().equals("fileUploadTransaction")) {
            Counter.Config configWithLabel = config.withLabels(new Label("transactionType", "fileUpload"));
            Counter counter = metrics.getOrCreate(config);
            counter.increment();
        } else if(transaction.getType().equals("fileDeleteTransaction")) {
            Counter.Config configWithLabel = config.withLabels(new Label("transactionType", "fileDelete"));
            Counter counter = metrics.getOrCreate(config);
            counter.increment();

>         }

Reviewer 4: This is not a realistic example. We have ~50 transaction types, and the number is increasing. A method with 50 if-else branches will hopefully never make it through code review. 🙂

Author: ok, do you have a better one? I just wanted to have any example
